### PR TITLE
K8SPXC-279 Add operator admin user

### PR DESCRIPTION
**Pull request by nonemax**
[K8SPXC-279](https://jira.percona.com/browse/K8SPXC-279)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | failed |
| scaling | failed |
| init-deploy | failed |
| storage | failed |
| big-data | failed |
| scaling-proxysql | failed |
| self-healing | failed |
| restore-to-encrypted-cluster | failed |
| limits | failed |
| monitoring | failed |
| upgrade | failed |
| demand-backup | failed |
| self-healing-advanced | failed |
| upgrade-consistency | failed |
| operator-self-healing | failed |
| scheduled-backup | failed |
| security-context | failed |
| monitoring-2-0 | failed |
| demand-backup-encrypted-with-tls | failed |
| affinity | failed |
| one-pod | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |

File: build/pxc-entrypoint.sh [hunk 1]
```diff
@@ -252,6 +252,10 @@ if [ -z "$CLUSTER_JOIN" ] && [ "$1" = 'mysqld' -a -z "$wantHelp" ]; then
 			ALTER USER 'root'@'localhost' IDENTIFIED BY '${MYSQL_ROOT_PASSWORD}' ;
 			GRANT ALL ON *.* TO 'root'@'localhost' WITH GRANT OPTION ;
 			${rootCreate}
+			REVOKE SYSTEM_USER ON *.* FROM root ;
+
+			CREATE USER 'operatoradmin'@'${MYSQL_ROOT_HOST}' IDENTIFIED BY '${OPERATOR_ADMIN_PASSWORD}' ;
```
Comment (delgod) [inline build/pxc-entrypoint.sh:257]: why not, just `operator`?
```suggestion
			CREATE USER 'operator'@'${MYSQL_ROOT_HOST}' IDENTIFIED BY '${OPERATOR_ADMIN_PASSWORD}' ;
```

File: pkg/pxc/users/users.go [hunk 2]
```diff
@@ -38,6 +38,39 @@ func NewManager(addr string, user, pass string) (Manager, error) {
...
 }
 
+func (u *Manager) CreateOperatorAdminUser(pass string) error {
+	defer u.db.Close()
+	tx, err := u.db.Begin()
+	if err != nil {
+		return errors.Wrap(err, "begin transaction")
+	}
+
+	_, err = tx.Exec("CREATE USER 'operatoradmin'@'%' IDENTIFIED BY ?", pass)
```
Comment (delgod) [inline pkg/pxc/users/users.go:48]: I think it is also needed to run `GRANT ALL ON *.* TO 'operatoradmin'@'${MYSQL_ROOT_HOST}' WITH GRANT OPTION ;`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| recreate | passed |
| scaling | failed |
| big-data | failed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| limits | passed |
| restore-to-encrypted-cluster | failed |
| monitoring | passed |
| upgrade | failed |
| demand-backup | passed |
| upgrade-consistency | passed |
| self-healing-advanced | failed |
| monitoring-2-0 | passed |
| security-context | passed |
| affinity | passed |
| one-pod | failed |
| auto-tuning | passed |
| operator-self-healing | failed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| demand-backup-encrypted-with-tls | passed |

File: pkg/pxc/users/users.go (hunk 2 above)
Comment (nonemax) [reply pkg/pxc/users/users.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/436/commits/8f0d3d72ea10026c22a82cce1e00528083f56ef7

File: pkg/controller/pxc/controller.go [hunk 3]
```diff
@@ -299,6 +299,11 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 		return reconcile.Result{}, fmt.Errorf("failed to ensure version: %v", err)
 	}
 
+	err = r.manageOperatorAdminUser(o)
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/controller/pxc/controller.go:302]: https://github.com/percona/percona-xtradb-cluster-operator/pull/436/commits/8f0d3d72ea10026c22a82cce1e00528083f56ef7

File: build/pxc-entrypoint.sh (hunk 1 above)
Comment (nonemax) [reply build/pxc-entrypoint.sh]: https://github.com/percona/percona-xtradb-cluster-operator/pull/436/commits/8f0d3d72ea10026c22a82cce1e00528083f56ef7

File: pkg/controller/pxc/secrets.go [hunk 4]
```diff
@@ -75,7 +79,8 @@ const (
 		"0123456789"
 )
 
-func generatePass() ([]byte, error) {
+//GeneratePass generate random password
+func GeneratePass() ([]byte, error) {
```
Comment (dAdAbird) [inline pkg/controller/pxc/secrets.go:83]: why is this method became exported?

Comment (nonemax) [reply pkg/controller/pxc/secrets.go]: fixed https://github.com/percona/percona-xtradb-cluster-operator/pull/436/commits/cd59bf0031fcfa410a6f073ef55301f1f0d3c536

File: pkg/controller/pxc/users.go [hunk 5]
```diff
@@ -309,6 +323,55 @@ func (r *ReconcilePerconaXtraDBCluster) getInternalSysUsersSecretObj(cr *api.Per
...
+	}
+	var exist bool
+	for name := range sysUsersSecretObj.Data {
+		switch name {
+		case "operator":
+			exist = true
+		}
+	}
+
+	if !exist {
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/436/commits/cd59bf0031fcfa410a6f073ef55301f1f0d3c536

File: pkg/pxc/users/users.go [hunk 6]
```diff
@@ -38,6 +38,48 @@ func NewManager(addr string, user, pass string) (Manager, error) {
 	return um, nil
 }
 
+func (u *Manager) CreateOperatorUser(pass string) error {
+	defer u.db.Close()
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/pxc/users/users.go]: added https://github.com/percona/percona-xtradb-cluster-operator/pull/436/commits/cd59bf0031fcfa410a6f073ef55301f1f0d3c536

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| scaling | passed |
| big-data | failed |
| storage | passed |
| recreate | passed |
| limits | passed |
| scaling-proxysql | passed |
| monitoring | passed |
| upgrade | failed |
| self-healing | failed |
| restore-to-encrypted-cluster | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| affinity | passed |
| demand-backup | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | failed |

File: pkg/controller/pxc/controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (delgod) [reply pkg/controller/pxc/controller.go:302]: now, we run it in both places, in reconcileUsers and here 😞

[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/controller/pxc/controller.go]: removed https://github.com/percona/percona-xtradb-cluster-operator/pull/436/commits/9d2144010ce24b9bbc720c4f04a23f1345b023b2

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| init-deploy | failed |
| big-data | failed |
| storage | passed |
| recreate | failed |
| limits | passed |
| self-healing | failed |
| self-healing-advanced | passed |
| restore-to-encrypted-cluster | passed |
| monitoring | passed |
| scaling-proxysql | passed |
| operator-self-healing | passed |
| upgrade | failed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| scheduled-backup | passed |
| one-pod | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| demand-backup-encrypted-with-tls | failed |
| upgrade-consistency | failed |

**Pull request merged**